### DISCUSSION TITLE
Server-Side Issue: Row Level security auth.uid() check doesn't work

### DISCUSSION THREAD

**Discussion by devhe4d** [supabase/supabase#6592] [Questions] [answered]
Hey everyone! hope you all have a good day.

I'm new to supabase, so my question might be a little noobish, I also searched a lot around this problem of mine but I couldn't find any proper solution or guide,

There is a table called **Projects** which I turned on the RLS for it with a rule for **All actions** which is **auth.uid() = user_id**

so the scenario is that every user should only read their own project not others, but the problem I'm facing is the **Select** command returns an empty array although I'm logged in with a authorised user.

I'm also using Supabase own user management system for authentication and everything works fine.

what could be the issue?

<img width="1153" alt="Screenshot 2022-04-27 at 19 56 12" src="https://user-images.githubusercontent.com/58355247/165590398-0484a265-80bf-4a2f-a5c7-79ec0250ac38.png">

this code returns an empty array although I'm logged in.
<img width="559" alt="Screenshot 2022-04-27 at 20 01 40" src="https://user-images.githubusercontent.com/58355247/165590500-99616c32-f20d-46be-b18f-51d5eacaf300.png">

**Comment by GaryAustin1** [comment]
Is user_id a uuid type?
If so then try setting policy to auth.role() = 'anon' and if that works then you are not logged in when you make the request for some reason.

**Reply by devhe4d** [reply]
> Is user_id a uuid type? If so then try setting policy to auth.role() = 'anon' and if that works then you are not logged in when you make the request for some reason.

it doesn't allow me to do that. it throws this error `Error updating policy: invalid input syntax for type uuid: "anon"`

and yes the user_id is uuid which Supabase automatically generate when you use its API to sign up users

**Reply by GaryAustin1** [reply]
You changed it to auth.ROLE() = 'anon'?  And single quotes around 'anon'?

**Reply by devhe4d** [reply]
> You changed it to auth.ROLE() = 'anon'? And single quotes around 'anon'?

sorry my mistake.
Now I changed RLS to `auth.role() = 'anon'` and I can receive data again. but the problem stands since now I'm receiving all the projects, not the ones owned by the current user.

this route ( where the request happens ) is protected, I'm checking it with ` supabase.auth.api.getUserByCookie()` on the server. 
so the current user is present and fully authorised when this request happens

**Reply by GaryAustin1** [reply]
The fact that it works with 'anon' means for whatever reason you do not have the user JWT token set in the client at the point you make that call.  So you can put back auth.uid() as the issue is not the RLS.

If the user token is not set in supabase when you make the call, it defaults to the anon token.

**Reply by devhe4d** [reply]
> The fact that it works with 'anon' means for whatever reason you do not have the user JWT token set in the client at the point you make that call. So you can put back auth.uid() as the issue is not the RLS.
> 
> If the user token is not set in supabase when you make the call, it defaults to the anon token.

but when I check for user right before this request with `supabase.auth.api.getUserByCookie(req)` on the server, it shows me the current user.

**Reply by GaryAustin1** [reply]
Try running this before you  make calls after you getUser...
const { user, error } = supabase.auth.setAuth(access_token)  where access token is from the user object (I think).
I don't use server side stuff myself.  If setAuth does not work or you can't find the answer searching discussions here then probably best to start a new question as your title will not get you server users for help.  You can also ask on github.  You will probably need to show your basic code either way.

**Reply by devhe4d** [reply]
Thank you, will check for these and see what happens.

**Comment by devhe4d** [comment]
Update

**Comment by devhe4d** [comment, marked as answer]
Solution:

since Server talks to Supabase anonymously via the `anon` key, you need to overwrite it with the current user `access_token` (which is a JWT). 

you can simply access token from the `req` object like this:

` const { token } = await supbase.auth.api.getUserByCookie(req)`
` supabase.auth.setAuth(token)`

**Comment by johnta0** [comment]
As of Apr 27, 2023, you're not going to bump into this problem as long as you use `@supabase/auth-helpers-nextjs` package. 
Ref: https://supabase.com/docs/guides/auth/auth-helpers/nextjs#server-side-data-fetching-with-rls

**Reply by nmerchant** [reply]
Are you sure about this?  I seem to be having this issue despite using latest `@supabase/auth-helpers-nextjs`.  If I set my RLS policy to 'true' I get all the rows back no issue, but if I set the policy to check for `auth.uid() = user_id` (user_id is a foreign key referencing the auth.users table), I get 0 records back:

```
import { createRouteHandlerClient } from '@supabase/auth-helpers-nextjs';
import { cookies } from 'next/headers';
import { NextResponse } from 'next/server';

export async function GET() {
        const supabase = createRouteHandlerClient({ cookies });
        
        const {
            data: { user },
        } = await supabase.auth.getUser();

        console.log(user); // this is null

        const { data: userData } = await supabase.from('users').select('*');

        const res = await fetch('someUrl', {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                'apiKey': process.env.my_api_key || ''
            }
        });
    
        const data = await res.json();
        return NextResponse.json({ data });
}
```

**Reply by Ricerob** [reply]
@nmerchant Any update on this? I'm running into this same issue. I'm new to RSC and NextJS 13 so I'm struggling trying to figure out what is using server methods and what's using client methods and everything in between.

**Reply by nmerchant** [reply]
> @nmerchant Any update on this? I'm running into this same issue. I'm new to RSC and NextJS 13 so I'm struggling trying to figure out what is using server methods and what's using client methods and everything in between.

Apologies, I did something to fix it but I didn't document well enough and now having a hard time remembering.  I THINK it had something to do with this change, give it a try and let me know if it works...

Change
```
const supabase = createServerComponentClient({ cookies });
```

to this:
```
const cookieStore = cookies();
const supabase = createServerComponentClient({ cookies: () => cookieStore });
```

**Reply by Ricerob** [reply]
@nmerchant No problem. I ended up needing the component to be a client component anyways - if I end up using a server component, I'll lyk. Thanks for the response to an older thread!

**Reply by nmerchant** [reply]
@Ricerob Sure thing.  Also, I believe this was relevant: you need to manually pass headers when using server components.
https://jonmeyers.io/blog/forwarding-cookies-from-server-components-to-route-handlers-with-next-js-app-router
